sylvia-derive/src/contract/communication/struct_msg.rs

```diff
@@ -32,138 +26,109 @@ impl<'a> StructMessage<'a> {
-                quote! {}
-            }
-        }
-    }
-
-    pub fn emit_struct(&self, name: &Ident) -> TokenStream {
         let sylvia = crate_module();
 
         let Self {
+            source,
             contract_type,
-            fields,
-            function_name,
+            variants,
             generics,
-            unused_generics,
-            wheres,
-            full_where,
-            result,
-            msg_attr,
+            error,
             custom,
             msg_attrs_to_forward,
         } = self;
 
-        let ctx_type = msg_attr
+        let Some(variant) = variants.get_only_variant() else {
+            return quote! {};
+        };
+
+        let used_generics = variants.used_generics();
+        let unused_generics = variants.unused_generics();
+        let full_where = &source.generics.where_clause;
+        let wheres = filter_wheres(full_where, generics, used_generics);
+        let where_clause = as_where_clause(&wheres);
+        let bracketed_used_generics = emit_bracketed_generics(used_generics);
+        let bracketed_unused_generics = emit_bracketed_generics(unused_generics);
+
+        let ret_type = variant
             .msg_type()
-            .emit_ctx_type(&custom.query_or_default());
-        let fields_names: Vec<_> = fields.iter().map(MsgField::name).collect();
-        let parameters = fields.iter().map(MsgField::emit_method_field);
-        let fields = fields.iter().map(MsgField::emit_pub);
-
-        let where_clause = as_where_clause(wheres);
-        let generics = emit_bracketed_generics(generics);
-        let unused_generics = emit_bracketed_generics(unused_generics);
+            .emit_result_type(&custom.msg_or_default(), &error.error);
+        let name = variant.msg_type().emit_msg_name();
+        let function_name = variant.function_name();
+        let mut msg_name = variant.msg_type().emit_msg_name();
+        msg_name.set_span(function_name.span());
+
+        let ctx_type = variant.msg_type().emit_ctx_type(&custom.query_or_default());
+        let fields_names: Vec<_> = variant.fields().iter().map(MsgField::name).collect();
+        let parameters = variant.fields().iter().map(MsgField::emit_method_field);
+        let fields = variant.fields().iter().map(MsgField::emit_pub);
+
         let msg_attrs_to_forward = msg_attrs_to_forward.iter().map(|attr| &attr.attrs);
 
         quote! {
             #[allow(clippy::derive_partial_eq_without_eq)]
             #[derive(#sylvia ::serde::Serialize, #sylvia ::serde::Deserialize, Clone, Debug, PartialEq, #sylvia ::schemars::JsonSchema)]
             #( #[ #msg_attrs_to_forward ] )*
             #[serde(rename_all="snake_case")]
-            pub struct #name #generics {
+            pub struct #name #bracketed_used_generics {
                 #(#fields,)*
             }
 
-            impl #generics #name #generics #where_clause {
+            impl #bracketed_used_generics #name #bracketed_used_generics #where_clause {
```

Comment:
The set of generics have been changed here - it'd be beneficial to see that change somewhere in the test cases

Reply:
It didn't. The previous naming was incorrect.

```rust
generics: used_generics,
```